File: apps/mobile/android/app/src/main/java/com/healthcoachai/app/config/ConfigManager.kt
package com.healthcoachai.app.config

import android.content.Context
import android.content.SharedPreferences
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.State
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.encodeToString
import kotlinx.serialization.decodeFromString
import java.net.URL
import java.util.Date
import java.util.concurrent.TimeUnit
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response

/**
 * Configuration Manager for HealthCoachAI Android App
 * Implements Phase 7 requirement: "Config consumption from backend; no secrets embedded"
 * All configuration loaded from backend API, no hardcoded secrets in client
 */
class ConfigManager private constructor(private val context: Context) {
    companion object {
        @Volatile
        private var INSTANCE: ConfigManager? = null
        
        fun getInstance(context: Context): ConfigManager {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: ConfigManager(context.applicationContext).also { INSTANCE = it }
            }
        }
    }
    
    private val prefs: SharedPreferences = context.getSharedPreferences("app_config", Context.MODE_PRIVATE)
    private val json = Json { ignoreUnknownKeys = true }
    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()
    
    private val _config = MutableStateFlow<AppConfig?>(null)
    val config: StateFlow<AppConfig?> = _config.asStateFlow()
    
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    
    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()
    
    private val configCacheKey = "app_config_cache"
    private val configCacheTimestampKey = "app_config_timestamp"
    private val cacheValidityHours = 24L // 24 hours
    
    init {
        loadCachedConfig()
    }
    
    // MARK: - Public Methods
    
    suspend fun loadConfiguration() {
        _isLoading.value = true
        _error.value = null
        
        try {
            val freshConfig = fetchConfigFromBackend()
            _config.value = freshConfig
            cacheConfig(freshConfig)
        } catch (e: Exception) {
            _error.value = e.message ?: "Unknown error occurred"
            // Fall back to cached config if available
            if (_config.value == null) {
                loadCachedConfig()
            }
        } finally {
            _isLoading.value = false
        }
    }
    
    fun getAPIBaseURL(): String {
        return _config.value?.apiBaseURL ?: "https://api.healthcoachai.com"
    }
    
    fun getFeatureFlags(): FeatureFlags {
        return _config.value?.featureFlags ?: FeatureFlags()
    }
    
    fun getUISettings(): UISettings {
        return _config.value?.uiSettings ?: UISettings()
    }
    
    // MARK: - Private Methods
    
    private fun loadCachedConfig() {
        val cachedData = prefs.getString(configCacheKey, null)
        val timestamp = prefs.getLong(configCacheTimestampKey, 0)
        
        if (cachedData != null && isConfigCacheValid(timestamp)) {
            try {
                val cachedConfig = json.decodeFromString<AppConfig>(cachedData)
                _config.value = cachedConfig
            } catch (e: Exception) {
                println("Failed to decode cached config: ${e.message}")
            }
        }
    }
    
    private fun cacheConfig(config: AppConfig) {
        try {
            val configJson = json.encodeToString(config)
            prefs.edit()
                .putString(configCacheKey, configJson)
                .putLong(configCacheTimestampKey, System.currentTimeMillis())
                .apply()
        } catch (e: Exception) {
            println("Failed to cache config: ${e.message}")
        }
    }
    
    private fun isConfigCacheValid(timestamp: Long): Boolean {
        val ageHours = (System.currentTimeMillis() - timestamp) / (1000 * 60 * 60)
        return ageHours < cacheValidityHours
    }
    
    private fun fetchConfigFromBackend(): AppConfig {
        // Environment-based backend URL (no secrets in code)
        val baseURL = System.getenv("API_BASE_URL") ?: "https://api.healthcoachai.com"
        val url = "$baseURL/api/mobile/config"
        
        val request = Request.Builder()
            .url(url)
            .get()
            .addHeader("Accept", "application/json")
            .addHeader("User-Agent", "HealthCoachAI-Android/1.0")
            .build()
        
        val response = httpClient.newCall(request).execute()
        
        if (!response.isSuccessful) {
            throw ConfigException("Server error: ${response.code}")
        }
        
        val responseBody = response.body?.string()
            ?: throw ConfigException("Empty response body")
        
        return json.decodeFromString<AppConfig>(responseBody)
    }
}

// MARK: - Data Models

@Serializable
data class AppConfig(
    @SerialName("api_base_url") val apiBaseURL: String,
    @SerialName("feature_flags") val featureFlags: FeatureFlags,
    @SerialName("ui_settings") val uiSettings: UISettings,
    val version: String,
    @SerialName("supported_versions") val supportedVersions: List<String>,
    @SerialName("last_updated") val lastUpdated: Long
)

@Serializable
data class FeatureFlags(
    @SerialName("enable_meal_planning") val enableMealPlanning: Boolean = true,
    @SerialName("enable_fitness_tracking") val enableFitnessTracking: Boolean = true,
    @SerialName("enable_ai_chat") val enableAIChat: Boolean = false,
    @SerialName("enable_social_features") val enableSocialFeatures: Boolean = false,
    @SerialName("enable_premium_features") val enablePremiumFeatures: Boolean = false,
    @SerialName("enable_hinglish_support") val enableHinglishSupport: Boolean = true,
    @SerialName("enable_offline_mode") val enableOfflineMode: Boolean = false
)

@Serializable
data class UISettings(
    @SerialName("refresh_interval_minutes") val refreshIntervalMinutes: Int = 30,
    @SerialName("max_cache_age_days") val maxCacheAgeDays: Int = 7,
    @SerialName("animation_duration_ms") val animationDurationMs: Int = 250,
    @SerialName("enable_animations") val enableAnimations: Boolean = true,
    @SerialName("default_language") val defaultLanguage: String = "en",
    @SerialName("supported_languages") val supportedLanguages: List<String> = listOf("en", "hi")
)

class ConfigException(message: String) : Exception(message)

File: apps/mobile/ios/HealthCoachAI/Config/ConfigManager.swift
import Foundation

/**
 * Configuration Manager for HealthCoachAI iOS App
 * Implements Phase 7 requirement: "Config consumption from backend; no secrets embedded"
 * All configuration loaded from backend API, no hardcoded secrets in client
 */
class ConfigManager: ObservableObject {
    static let shared = ConfigManager()
    
    @Published var config: AppConfig?
    @Published var isLoading = false
    @Published var error: Error?
    
    private let userDefaults = UserDefaults.standard
    private let configCacheKey = "app_config_cache"
    private let configCacheTimestampKey = "app_config_timestamp"
    private let cacheValidityHours: TimeInterval = 24 // 24 hours
    
    private init() {
        loadCachedConfig()
    }
    
    // MARK: - Public Methods
    
    func loadConfiguration() async {
        await MainActor.run {
            isLoading = true
            error = nil
        }
        
        do {
            let freshConfig = try await fetchConfigFromBackend()
            await MainActor.run {
                self.config = freshConfig
                self.isLoading = false
                self.cacheConfig(freshConfig)
            }
        } catch {
            await MainActor.run {
                self.error = error
                self.isLoading = false
                // Fall back to cached config if available
                if self.config == nil {
                    self.loadCachedConfig()
                }
            }
        }
    }
    
    func getAPIBaseURL() -> String {
        return config?.apiBaseURL ?? "https://api.healthcoachai.com" // Default fallback
    }
    
    func getFeatureFlags() -> FeatureFlags {
        return config?.featureFlags ?? FeatureFlags()
    }
    
    func getUISettings() -> UISettings {
        return config?.uiSettings ?? UISettings()
    }
    
    // MARK: - Private Methods
    
    private func loadCachedConfig() {
        guard let data = userDefaults.data(forKey: configCacheKey),
              let timestamp = userDefaults.object(forKey: configCacheTimestampKey) as? Date,
              Date().timeIntervalSince(timestamp) < cacheValidityHours * 3600 else {
            return
        }
        
        do {
            let cachedConfig = try JSONDecoder().decode(AppConfig.self, from: data)
            self.config = cachedConfig
        } catch {
            print("Failed to decode cached config: \(error)")
        }
    }
    
    private func cacheConfig(_ config: AppConfig) {
        do {
            let data = try JSONEncoder().encode(config)
            userDefaults.set(data, forKey: configCacheKey)
            userDefaults.set(Date(), forKey: configCacheTimestampKey)
        } catch {
            print("Failed to cache config: \(error)")
        }
    }
    
    private func fetchConfigFromBackend() async throws -> AppConfig {
        // Environment-based backend URL (no secrets in code)
        let baseURL = ProcessInfo.processInfo.environment["API_BASE_URL"] ?? "https://api.healthcoachai.com"
        
        guard let url = URL(string: "\(baseURL)/api/mobile/config") else {
            throw ConfigError.invalidURL
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("HealthCoachAI-iOS/1.0", forHTTPHeaderField: "User-Agent")
        
        let (data, response) = try await URLSession.shared.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse,
              200...299 ~= httpResponse.statusCode else {
            throw ConfigError.serverError
        }
        
        let config = try JSONDecoder().decode(AppConfig.self, from: data)
        return config
    }
}

// MARK: - Data Models

struct AppConfig: Codable {
    let apiBaseURL: String
    let featureFlags: FeatureFlags
    let uiSettings: UISettings
    let version: String
    let supportedVersions: [String]
    let lastUpdated: Date
    
    enum CodingKeys: String, CodingKey {
        case apiBaseURL = "api_base_url"
        case featureFlags = "feature_flags"
        case uiSettings = "ui_settings"
        case version
        case supportedVersions = "supported_versions"
        case lastUpdated = "last_updated"
    }
}

struct FeatureFlags: Codable {
    let enableMealPlanning: Bool
    let enableFitnessTracking: Bool
    let enableAIChat: Bool
    let enableSocialFeatures: Bool
    let enablePremiumFeatures: Bool
    let enableHinglishSupport: Bool
    let enableOfflineMode: Bool
    
    init() {
        // Default values when no config available
        self.enableMealPlanning = true
        self.enableFitnessTracking = true
        self.enableAIChat = false
        self.enableSocialFeatures = false
        self.enablePremiumFeatures = false
        self.enableHinglishSupport = true
        self.enableOfflineMode = false
    }
    
    enum CodingKeys: String, CodingKey {
        case enableMealPlanning = "enable_meal_planning"
        case enableFitnessTracking = "enable_fitness_tracking"
        case enableAIChat = "enable_ai_chat"
        case enableSocialFeatures = "enable_social_features"
        case enablePremiumFeatures = "enable_premium_features"
        case enableHinglishSupport = "enable_hinglish_support"
        case enableOfflineMode = "enable_offline_mode"
    }
}

struct UISettings: Codable {
    let refreshIntervalMinutes: Int
    let maxCacheAgeDays: Int
    let animationDurationMs: Int
    let enableAnimations: Bool
    let defaultLanguage: String
    let supportedLanguages: [String]
    
    init() {
        // Default values
        self.refreshIntervalMinutes = 30
        self.maxCacheAgeDays = 7
        self.animationDurationMs = 250
        self.enableAnimations = true
        self.defaultLanguage = "en"
        self.supportedLanguages = ["en", "hi"]
    }
    
    enum CodingKeys: String, CodingKey {
        case refreshIntervalMinutes = "refresh_interval_minutes"
        case maxCacheAgeDays = "max_cache_age_days"
        case animationDurationMs = "animation_duration_ms"
        case enableAnimations = "enable_animations"
        case defaultLanguage = "default_language"
        case supportedLanguages = "supported_languages"
    }
}

enum ConfigError: Error, LocalizedError {
    case invalidURL
    case serverError
    case decodingError
    
    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid configuration URL"
        case .serverError:
            return "Server error while fetching configuration"
        case .decodingError:
            return "Failed to decode configuration data"
        }
    }
}